python/redmonster/physics/zfinder.py

```diff
@@ -63,198 +130,176 @@ def read_template(self):
-                                        self.chi2_null[i]
+                # multiprocessing
+
+                func_args = []
+                for j in range(self.templates_flat.shape[0]):
+                    if self.npoly>0 :
+                        arguments = {"j":j,"poly_fft":poly_fft[i], "t_fft":self.t_fft[j], "t2_fft":self.t2_fft[j], "data_fft":data_fft[i], "ivar_fft":ivar_fft[i], "pmat_pol":pmat, "bvec_pol":bvec, "chi2_0":self.sn2_data[i], "chi2_null":self.chi2_null[i],"num_z":num_z, "npixstep":self.npixstep, "zminpix":zminpix,"flag_val_neg_model":flag_val_neg_model}
+                    else :
+                        arguments = {"j":j,"t_fft":self.t_fft[j], "t2_fft":self.t2_fft[j], "data_fft":data_fft[i], "ivar_fft":ivar_fft[i], "chi2_0":self.sn2_data[i], "num_z":num_z, "npixstep":self.npixstep, "zminpix":zminpix, "flag_val_neg_model":flag_val_neg_model}
+                    func_args.append(arguments)
+
+                results = None
+                if self.nproc > 1:
+                    pool = multiprocessing.Pool(self.nproc)
+                    if self.npoly>0 :
+                        results = pool.map(_zchi2, func_args)
+                    else :
+                        results = pool.map(_zchi2_no_poly, func_args)
+                    pool.close()
+                    pool.join()
+                else:
+                    if self.npoly>0 :
+                        results = [ _zchi2(x) for x in func_args ]
+                    else :
+                        results = [ _zchi2_no_poly(x) for x in func_args ]
+
+                for result in results :
+                    j                  = result[0]
+                    zchi2arr[i,j]      = result[1]
+                    temp_zwarning[i,j] = result[2]
+
+                stop=time.time()
+
+                print("INFO fitted fiber %d/%d, chi2_null=%f, %d templates in %s, npoly=%d, using %d procs in %f sec"%(i+1, specs.shape[0],self.chi2_null[i],self.templates_flat.shape[0],self.fname,self.npoly,self.nproc,stop-start))
+
+
         # Use only neg_model flag from best fit model/redshift and add
         # it to self.zwarning
-        for i in xrange(self.zwarning.shape[0]):
+        for i in range(self.zwarning.shape[0]):
             minpos = ( n.where(zchi2arr[i] == n.min(zchi2arr[i]))[0][0],
                       n.where(zchi2arr[i] == n.min(zchi2arr[i]))[1][0] )
             self.zwarning[i] = int(self.zwarning[i]) | \
                     int(temp_zwarning[i,minpos[0],minpos[1]])
         zchi2arr = n.reshape(zchi2arr, (specs.shape[0],) + self.origshape[:-1] +
                              (num_z,) )
         bestl = n.where(zchi2arr == n.min(zchi2arr))[-1][0]
-        if bounds_set:
-            thisz = ((10**(specloglam[0]))/self.tempwave[bestl+zminpix])-1
-        else:
-            thisz = ((10**(specloglam[0]))/self.tempwave[bestl])-1
+        thisz = ((10**(specloglam[0]))/self.tempwave[bestl+zminpix])-1
```

Comment:
It looks like this will still work, as the same logic is still there, but was moved elsewhere.  I don't understand why, though.